zcash_client_backend/src/sync.rs

```diff
@@ -0,0 +1,399 @@
+
+            Ok(meta)
+        })
+        .try_collect::<Vec<_>>()
+        .await?;
+
+    db_cache.write_block_metadata(&block_meta)?;
+
+    Ok(block_meta)
+}
+
+fn delete_cached_blocks(fsblockdb_root: &Path, block_meta: Vec<BlockMeta>) -> JoinHandle<()> {
+    let fsblockdb_root = fsblockdb_root.to_owned();
+    tokio::spawn(async move {
+        for meta in block_meta {
+            if let Err(e) = tokio::fs::remove_file(get_block_path(&fsblockdb_root, &meta)).await {
+                error!("Failed to remove {:?}: {}", meta, e);
+            }
+        }
+    })
+}
+
+/// Scans the given block range and checks for scanning errors that indicate the wallet's
+/// chain tip is out of sync with blockchain history.
+///
+/// Returns `true` if scanning these blocks materially changed the suggested scan ranges.
+fn scan_blocks<P, DbT>(
+    params: &P,
+    fsblockdb_root: &Path,
+    db_cache: &mut FsBlockDb,
+    db_data: &mut DbT,
+    scan_range: &ScanRange,
+) -> Result<bool, anyhow::Error>
+where
+    P: Parameters + Send + 'static,
+    DbT: WalletWrite,
+    DbT::Error: std::error::Error + Send + Sync + 'static,
+{
+    info!("Scanning {}", scan_range);
+    let scan_result = scan_cached_blocks(
+        params,
+        db_cache,
+        db_data,
+        scan_range.block_range().start,
+        scan_range.len(),
+    );
+
+    match scan_result {
+        Err(ChainError::Scan(err)) if err.is_continuity_error() => {
+            // Pick a height to rewind to, which must be at least one block before the
+            // height at which the error occurred, but may be an earlier height determined
+            // based on heuristics such as the platform, available bandwidth, size of
+            // recent CompactBlocks, etc.
+            let rewind_height = err.at_height().saturating_sub(10);
+            info!(
+                "Chain reorg detected at {}, rewinding to {}",
+                err.at_height(),
+                rewind_height,
+            );
+
+            // Rewind to the chosen height.
+            db_data.truncate_to_height(rewind_height)?;
+
+            // Delete cached blocks from rewind_height onwards.
+            //
+            // This does imply that assumed-valid blocks will be re-downloaded, but it is
+            // also possible that in the intervening time, a chain reorg has occurred that
+            // orphaned some of those blocks.
+            db_cache
+                .with_blocks(Some(rewind_height + 1), None, |block| {
+                    let meta = BlockMeta {
+                        height: block.height(),
+                        block_hash: block.hash(),
+                        block_time: block.time,
+                        // These values don't matter for deletion.
+                        sapling_outputs_count: 0,
+                        orchard_actions_count: 0,
+                    };
+                    std::fs::remove_file(get_block_path(fsblockdb_root, &meta))
```

Comment:
why do we individually remove files instead of relying on truncate_to_height?

Reply:
`WalletWrite::truncate_to_height` is truncating the _wallet database_; it has no knowledge about the internal details of the block cache. This file deletion is part of the concrete behaviour required to use `FsBlockDb`, and as such it should move behind the new trait (i.e. there should be an equivalent `BlockCache::truncate_to_height` method).